### DISCUSSION TITLE
How to do a depth-first search with recursive CTEs?

### DISCUSSION THREAD

**Discussion by LennartH** [duckdb/duckdb#15386] [Q&A] [answered]
Sometimes it's necessary/beneficial to process some data depth-first and not just order the results afterwards. What would be the easiest/cleanest way to do this in DuckDB?

SQLite allows [using an `ORDER BY` clause to control this](https://www.sqlite.org/lang_with.html#controlling_depth_first_versus_breadth_first_search_of_a_tree_using_order_by), is there something similar in DuckDB?

Postgres has [`SEARCH BREADTH/DEPTH FIRST BY`](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-SEARCH), but this is just convenience to add a column to order by afterwards and doesn't actually control the evaluation order.

**Comment by szarnyasg** [comment]
Great question. I don't think this is currently possible but maybe @kryonix knows some tricks.

**Comment by LennartH** [comment]
One way I could think of would be to determine the execution order beforehand and adding something like a row number to each record. In the recursive CTE have some kind of iteration counter and join the auxiliary table using that counter, twisting the recursive CTE into a "for .. in" loop. But that feels a bit dirty and if the executions have some kind of relation to each other you would have to store the entire state in one record and drag it through the whole loop, since you can only access the n-1 state and not the n-2 and "earlier" state.

To give a bit more context, I'm doing this years Advent of Code with DuckDB and one puzzle was basically to find all components in a highly connected graph (~20k nodes, ~60k edges). Doing it breadth-first exploded and my brain just refused to tell me how to do it depth-first. In the end I implemented a simplified algorithm from [this paper](https://arxiv.org/pdf/1802.09478), but there has to be a reasonable depth-first implementation.

I might be wrong, but doing something depth-first should also allow the use of memoization (if previous results are accessible without blowing up space requirements for the state management).

I'm sure there are more (probably even real world) applications for depth-first searches, but this is mostly to satisfy my curiosity.

**Comment by kryonix** [comment, marked as answer]
First of all, I highly appreciate your effort trying to solve this years Advent of Code with DuckDB. We (myself and my colleagues at University of Tübingen) did the same thing for AoC 2022: https://github.com/DBatUTuebingen/Advent_of_Code/tree/main/2022 Maybe you can find a few ideas there.

> Great question. I don't think this is currently possible but maybe @kryonix knows some tricks.

As recursive CTEs are in fact _Turing-complete_ **everything** is possible—in theory 😉

> Postgres has [SEARCH BREADTH/DEPTH FIRST BY](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-SEARCH), but this is just convenience to add a column to order by afterwards and doesn't actually control the evaluation order.

The code your link navigates to actually shows the exact right idea. In order to transform a breadth-first search into a depth-first one, you essentially have to use a stack—As we do not have that in SQL, we can use an array to emulate one.

> SQLite allows using an ORDER BY clause to control this, is there something similar in DuckDB?

No, there is nothing similar in DuckDB. I'm not sure how SQLite uses the `ORDER BY` clause to control the evaluation order.

---

My example may be a bit over-engineered, but it provides hooks for further extensions.

```sql
CREATE TABLE edges(
  a INT,
  b INT
);

INSERT INTO edges
VALUES
  (1, 2),
  (1, 3),
  (2, 4),
  (4, 5),
  (4, 6);

-- Encodes the following tree:
-- 1
-- | \
-- 2  3
-- |
-- 4
-- | \
-- 5  6

WITH RECURSIVE breadth_first_traversal(node, path) AS (
  SELECT 1 AS node, [] :: STRUCT("from" INT, "to" INT)[] AS path -- Start with node 1 (root)
  UNION ALL
  SELECT e.b, [{'from': bft.node, 'to': e.b}] || bft.path
  FROM   breadth_first_traversal AS bft,
         edges AS e
  WHERE  bft.node = e.a
)
SELECT * FROM breadth_first_traversal;
```

This is your general run-of-the-mill breadth-first search. Nothing fancy going on.
We just start arbitrarily with node 1 and then keep joining with the edges table until we have no more edges to traverse.

To transform this into a depth-first search, we need to keep track of the siblings of a node. 

```sql
WITH RECURSIVE depth_first_traversal(stack, path) AS (
  SELECT [1] AS stack, [] :: STRUCT("from" INT, "to" INT)[] AS path -- Start with node 1 (root)
    UNION ALL
  (WITH siblings AS (
    SELECT ARRAY_AGG(e.b ORDER BY e.b ASC) AS siblings
    --                                ^^^
    -- This determines the order of traversal of the siblings
    FROM   depth_first_traversal AS dft,
           edges AS e
    WHERE  dft.stack[1] = e.a
  )
  SELECT x.*
  FROM   siblings AS _(s), LATERAL
         (SELECT s || dft.stack[2:] AS stack, -- Push the stack
                 [{'from': dft.stack[1], 'to': s[1]}] || dft.path AS path -- Add the edge to the path
          FROM   depth_first_traversal AS dft
          WHERE  array_length(s) > 0 -- There are more siblings to traverse
            UNION ALL
          SELECT dft.stack[2:], -- Pop the stack
                 [] :: STRUCT("from" INT, "to" INT)[] AS path -- Reset the path
          FROM   depth_first_traversal AS dft
          WHERE  array_length(s) IS NULL AND dft.stack <> [] -- No more siblings to traverse
          ) AS x
  )
)
SELECT * FROM depth_first_traversal;
```

This can get arbitrarily complex. I use this technique in our publication "Functional Programming on Top of SQL Engines", which you can find here:
https://db.cs.uni-tuebingen.de/publications/2022/functional-programming-on-top-of-sql-engines/cps-udf.pdf
Or in form of the second part of my dissertation:
https://db.cs.uni-tuebingen.de/publications/2024/new-compilation-methods-for-complex-user-defined-functions/dissertation_hirn_denis.pdf

We use it to actually express functional code in continuation-passing style in SQL. This also requires a stack to keep track of the continuation.
Same principle as above, just a bit more complex.

I also have a repository with the code for the publication here:
https://github.com/FP-on-Top-of-SQL-Engines/Code

---

Now, about memoization. This is a bit tricky. You would have to encode "memoization" rows into the working table. As you rightfully pointed out, this potentially explodes the number of rows, IFF there are not many opportunities to use the memoization rows.

Again, in the aforementioned publication, we show how to use memoization in a recursive CTE.
However, in the shown version, we do not employ memoization during one run of the recursive CTE.
We use memoization to store the results of previous runs of the recursive CTE.
But memoization during one run of the recursive CTE is also possible.

**Reply by LennartH** [reply]
Thank you so much for the detailed answer! This is more than I have hoped for and will definitely be enough for me to get a grip on how to construct depth-first queries.

I think the missing key for me was the integration of a stack and how to maintain it (the push/pop sections). Again, this is extremely valuable for me!

Is there another reason for resetting the path besides making the example easier to understand? Because this way all "context" is lost once a leaf has been encountered. For example, if the tree is extended into this:
```sql
-- 1
-- | \
-- 2  3
-- |
-- 4
-- | \
-- 5  6
--    |
--    7
--    |
--    8
INSERT INTO edges VALUES (6, 7), (7, 8);
```

After processing 5 the next node 6 is continued with an empty path that would be `4 - 2 - 1` with BFS. DuckDB output:
```
┌───────────┬────────────────────────────────────────────────────────────────────┐
│   stack   │                                path                                │
│  int32[]  │               struct("from" integer, "to" integer)[]               │
├───────────┼────────────────────────────────────────────────────────────────────┤
│ [1]       │ []                                                                 │
│ [2, 3]    │ [{'from': 1, 'to': 2}]                                             │
│ [4, 3]    │ [{'from': 2, 'to': 4}, {'from': 1, 'to': 2}]                       │
│ [5, 6, 3] │ [{'from': 4, 'to': 5}, {'from': 2, 'to': 4}, {'from': 1, 'to': 2}] │
│ [6, 3]    │ []                                                                 │
│ [7, 3]    │ [{'from': 6, 'to': 7}]                                             │
│ [8, 3]    │ [{'from': 7, 'to': 8}, {'from': 6, 'to': 7}]                       │
│ [3]       │ []                                                                 │
│ []        │ []                                                                 │
└───────────┴────────────────────────────────────────────────────────────────────┘
```

Of course continuing with the same path isn't correct either, because 5 can't be in the path of 6. So you probably would have to some information into the stack to know which part of the path to keep (the path length at time of insertion?).

Anyway, this is a great starter for all depth-first needs :ok_hand: 

----

As a side note, 2022 was the first AoC I participated in, so I will definitely take a look at your solution when I have some time.

**Reply by kryonix** [reply]
You are welcome! Feel free to reach out if you get stuck elsewhere. Recursive CTEs---and complex computation using SQL is our bread and butter over here 😉

> Is there another reason for resetting the path besides making the example easier to understand? Because this way all "context" is lost once a leaf has been encountered.

No, that choice is a completely arbitrary. Maybe you would want to (1) do some calculation at this part, (2) continue collecting the path, or (3) do something else. That is entirely up to you and what you are trying to achieve. I just wanted to give you a concrete example of a DFS in SQL. You do not even have to keep the general structure, the key really is how to use the _"stack"_.
